File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int INF = amount+1;
        vector<int> dp(amount+1,INF);
        dp[0] = 0;
        for (int n=1; n<=amount; n++){
            for (int c : coins){
                if (n-c >= 0)
                    dp[n] = min(dp[n],dp[n-c]+1);
            }
        }
        return dp[amount] == INF ? -1 : dp[amount];
    }
};

File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        INF = amount+1
        dp = [INF]*(amount+1)
        dp[0] = 0
        for k in range(1,amount+1):
            for denom in coins:
                if k - denom >= 0:
                    dp[k] = min(dp[k],dp[k-denom]+1)
        return -1 if dp[amount] == INF else dp[amount]